easypost/requestor.py

```diff
@@ -0,0 +1,193 @@
+            raise Error(
+                "No API key provided. Set an API key via \"easypost.api_key = 'APIKEY'. "
+                "Your API keys can be found in your EasyPost dashboard, or you can email us "
+                f"at {SUPPORT_EMAIL} for assistance."
+            )
+
+        abs_url = "%s%s" % (api_base, url or "")
+        params = self._objects_to_ids(params)
+
+        ua = {
+            "client_version": VERSION,
+            "lang": "python",
+            "publisher": "easypost",
+            "request_lib": request_lib,
+        }
+        for attr, func in (
+            ("lang_version", platform.python_version),
+            ("platform", platform.platform),
+            ("uname", lambda: " ".join(platform.uname())),
+        ):
+            try:
+                val = func()
+            except Exception as e:
+                val = "!! %s" % e
+            ua[attr] = val
+
+        if hasattr(ssl, "OPENSSL_VERSION"):
+            ua["openssl_version"] = ssl.OPENSSL_VERSION
+
+        headers = {
+            "X-Client-User-Agent": json.dumps(ua),
+            "User-Agent": USER_AGENT,
+            "Authorization": "Bearer %s" % my_api_key,
+            "Content-type": "application/json",
+        }
+
+        if request_lib == "urlfetch":
+            http_body, http_status = self.urlfetch_request(method, abs_url, headers, params)
+        elif request_lib == "requests":
+            http_body, http_status = self.requests_request(method, abs_url, headers, params)
+        else:
+            raise Error(f"Bug discovered: invalid request_lib: {request_lib}. Please report to {SUPPORT_EMAIL}.")
+
+        return http_body, http_status, my_api_key
+
+    def interpret_response(self, http_body, http_status):
+        try:
+            response = json.loads(http_body)
+        except Exception:
+            raise Error("Invalid response body from API: (%d) %s " % (http_status, http_body), http_status, http_body)
+        if not (200 <= http_status < 300):
+            self.handle_api_error(http_status, http_body, response)
+        return response
+
+    def requests_request(self, method, abs_url, headers, params):
+        # Importing here to avoid circular imports
+        from easypost import SUPPORT_EMAIL, requests_session, timeout
+
+        method = method.lower()
+        if method == "get" or method == "delete":
+            abs_url = self.add_params_to_url(abs_url, params)
+            data = None
+        elif method == "post" or method == "put":
+            data = json.dumps(params, default=self._utf8)
+        else:
+            raise Error(f"Bug discovered: invalid request method: {method}. Please report to {SUPPORT_EMAIL}.")
+
+        try:
+            result = requests_session.request(
+                method,
+                abs_url,
+                headers=headers,
+                data=data,
+                timeout=timeout,
+                verify=True,
+            )
```

Comment:
This is a kinda janky way of making requests for a few reasons (which you may not care about)...

1. The HTTP verb (`get`/`delete` vs `post`/`put`) has no true bearing on whether parameters should be assigned as URL query arguments or put in the message body. Gets can have json bodies, and posts can have URL parameters. This code is centrally mandating a relationship between them that is only true anecdotally, not logically.
2. Setting a `Content-Type` header to `application/json` when there's no message body doesn't hurt but it also isn't useful.
3. The requests library after version 2.4.2 has a mechanism already for setting the right header and doing the right thing for json messages by assigning the raw object to the `json` argument instead of assigning a string dump of the object to `data`, and the only reason this code doesn't use the more modern json mechanism is because of some decision from historic Python2 to accept raw bytes for addresses instead of strings and coerce them here with _utf8 instead of elsewhere. But all strings are unicode now in Python3 anyway and there's no reason for anyone to pass addresses in as raw bytes.
4. The requests library will also put your URL params in the right place for you without needing to manually call urlencode and concatenate strings by assigning to the `params` argument instead of to `data`.

Reply:
Thanks for the suggestion! For this PR we would like to keep it small and just shuffle around the code into separate modules. We can add an internal ticket to take a look into this and add another PR for it.